runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java

```diff
@@ -263,4 +284,81 @@ private static void applyLatencyTrackingInterval(
     long latencyTrackingInterval = options.getLatencyTrackingInterval();
     config.setLatencyTrackingInterval(latencyTrackingInterval);
   }
+
+  /**
+   * TODO: https://issues.apache.org/jira/browse/BEAM-5396
+   *
+   * <p>Remote stream environment that supports job execution with restore from savepoint. This
+   * class can be removed once Flink provides this functionality.
+   */
+  private static class BeamFlinkRemoteStreamEnvironment extends RemoteStreamEnvironment {
+    private final SavepointRestoreSettings restoreSettings;
+
+    public BeamFlinkRemoteStreamEnvironment(
+        String host,
+        int port,
+        Configuration clientConfiguration,
+        SavepointRestoreSettings restoreSettings,
+        String... jarFiles) {
+      super(host, port, clientConfiguration, jarFiles, null);
+      this.restoreSettings = restoreSettings;
+    }
+
+    // copied from RemoteStreamEnvironment and augmented to pass savepoint restore settings
+    @Override
+    protected JobExecutionResult executeRemotely(StreamGraph streamGraph, List<URL> jarFiles)
+        throws ProgramInvocationException {
+
+      List<URL> globalClasspaths = Collections.emptyList();
+      String host = super.getHost();
+      int port = super.getPort();
+
+      if (LOG.isInfoEnabled()) {
+        LOG.info("Running remotely at {}:{}", host, port);
+      }
+
+      ClassLoader usercodeClassLoader =
+          JobWithJars.buildUserCodeClassLoader(
+              jarFiles, globalClasspaths, getClass().getClassLoader());
+
+      Configuration configuration = new Configuration();
+      configuration.addAll(super.getClientConfiguration());
+
+      configuration.setString(JobManagerOptions.ADDRESS, host);
+      configuration.setInteger(JobManagerOptions.PORT, port);
+
+      configuration.setInteger(RestOptions.PORT, port);
+
+      final ClusterClient<?> client;
+      try {
+        if (CoreOptions.LEGACY_MODE.equals(configuration.getString(CoreOptions.MODE))) {
+          client = new StandaloneClusterClient(configuration);
+        } else {
+          client = new RestClusterClient<>(configuration, "RemoteStreamEnvironment");
+        }
+      } catch (Exception e) {
+        throw new ProgramInvocationException(
+            "Cannot establish connection to JobManager: " + e.getMessage(), e);
+      }
+
+      client.setPrintStatusDuringExecution(getConfig().isSysoutLoggingEnabled());
+
+      try {
+        return client
+            .run(streamGraph, jarFiles, globalClasspaths, usercodeClassLoader, restoreSettings)
+            .getJobExecutionResult();
+      } catch (ProgramInvocationException e) {
+        throw e;
+      } catch (Exception e) {
+        String term = e.getMessage() == null ? "." : (": " + e.getMessage());
+        throw new ProgramInvocationException("The program execution failed" + term, e);
```

Comment:
Not sure about the `term` string. I think we can drop it.

Reply:
This is unchanged copied, intentionally. I want to keep it as is, and in the future replace with utility that lives in Flink.